### ISSUE TITLE
Steps taken by the network layer

### ISSUE THREAD

**Issue by saum-g** [QTM3x/Quantum-Internet#57]
I think we should include the steps we expect the network layer to follow in the README of the network layer folder. This should include details such as what interface the network layer will provide to the upper layers, what interface it can expect from the lower layers and what information will be exclusive to the network layer.

It is reasonable to assume that the transport layer specifies the nodes which it needs to share an EPR pair and the minimum fidelity of the shared pair. It is then the job of the network layer to find the shortest time to make such a link. We can follow the steps mentioned in this [paper](https://arxiv.org/pdf/1907.11630.pdf) to achieve this exact purpose:

1. **Path Discovery**: Deciding the chain of nodes which connect the source and destination node.
2. **Entanglement Reservation**: Every pair of consecutive nodes in the chain which lack a shared entangled pair generate such a pair
3. **Entanglement Distribution**: Perform entanglement swap at every intermediate node in the chain to finally end up with an entangled pair between source and destination.

The steps 2 and 3 can also be combined into the "Prepare and Swap" protocol which was arrived upon in issue #19 and is also discussed in this [paper](https://arxiv.org/pdf/2005.14304.pdf). 

We can also include that, the network layer is free to generate more number of lower fidelity entangled pairs if it reduces the latency. This is not discussed in the first paper.

**Comment by BassemSafieldeen** [comment]
> I think we should include the steps we expect the network layer to follow in the README of the network layer folder. This should include details such as what interface the network layer will provide to the upper layers, what interface it can expect from the lower layers and what information will be exclusive to the network layer.
> 

I agree. Please do it. But I suggest you have a look at the code of the network layer on the development branch first -- see how it currently interacts with the other layers to have a better idea of what we have and what is missing. You can even run the code now: Run the cells in the BB84.ipynb in  --- Simulations --- and watch the printed messages.

> It is reasonable to assume that the transport layer specifies the nodes which it needs to share an EPR pair and the minimum fidelity of the shared pair. It is then the job of the network layer to find the shortest time to make such a link. We can follow the steps mentioned in this [paper](https://arxiv.org/pdf/1907.11630.pdf) to achieve this exact purpose:
> 
> 1. **Path Discovery**: Deciding the chain of nodes which connect the source and destination node.
> 2. **Entanglement Reservation**: Every pair of consecutive nodes in the chain which lack a shared entangled pair generate such a pair
> 3. **Entanglement Distribution**: Perform entanglement swap at every intermediate node in the chain to finally end up with an entangled pair between source and destination.
> 
> The steps 2 and 3 can also be combined into the "Prepare and Swap" protocol which was arrived upon in issue #19 and is also discussed in this [paper](https://arxiv.org/pdf/2005.14304.pdf).
> 
> We can also include that, the network layer is free to generate more number of lower fidelity entangled pairs if it reduces the latency. This is not discussed in the first paper.

Also see how the network layer currently interacts with the protocol ExampleProtocol1.
